shared/js/background/classes/tab-state.js

```diff
@@ -0,0 +1,177 @@
+        this.adClick = null
+        /** @type Record<string, import('./tracker').Tracker> */
+        this.trackers = {}
+        this.referrer = null
+
+        this.allowlisted = false // user-allowlisted sites; applies to all privacy features
+        this.allowlistOptIn = false
+        this.denylisted = false
+        // Whilst restoring, prevent the tab data being stored
+        if (!restoring) {
+            Storage.backup(this)
+        }
+    }
+
+    static async done () {
+        await Storage.done()
+    }
+
+    /**
+     * @template {InstanceType<typeof TabState>} T
+     * @template {keyof T} K
+     * @param {K} key
+     * @param {T[K]} value
+     */
+    setValue (key, value) {
+        // @ts-expect-error - we know this is a valid key, ts doesn't seem to understand T matches this
+        this[key] = value
+        // Fire and forget storage backup
+        Storage.backup(this)
+    }
+
+    /**
+     * Restores a tab state from storage.
+     * @param {number} tabId
+     * @returns {Promise<TabState | null>}
+     */
+    static async restore (tabId) {
+        const data = await Storage.get(tabId)
+        if (!data) {
+            return null
+        }
+        const parsedData = JSON.parse(data)
+        const state = new TabState(parsedData, true)
+        for (const key of Object.keys(parsedData)) {
+            state[key] = parsedData[key]
+        }
+        Storage.backup(state)
+        return state
+    }
+
+    /**
+     * Used for removing the stored tab state.
+     * @param {number} tabId
+     */
+    static async delete (tabId) {
+        await Storage.delete(tabId)
+    }
+}
+
+/**
+ * Singleton that handles storage of tab state to session storage.
+ * Guarantees that the tasks are performed in the order they are added.
+ */
+class StorageInstance {
+    taskQueue = []
+    processing = false
+
+    /**
+     * Awaits until the storage queue is empty.
+     * @returns {Promise<void>}
+     */
+    async done () {
+        const queue = this.taskQueue
+        await Promise.all(queue)
+    }
+
+    /**
+     * Adds a task to the storage queue, prevents tasks from being executed in parallel.
+     * Returns the result of the task.
+     * @template T
+     * @param {() => Promise<T>} task
+     * @returns {Promise<T>}
+     */
+    async addTask (task) {
+        let done = _ => {}
+        this.taskQueue.push(async () => {
+            const value = await Promise.resolve(task())
+            done(value)
+        })
+        this.processQueue()
+        return new Promise(resolve => {
+            done = resolve
+        })
+    }
+
+    /**
+     * Processes the storage queue in order.
+     */
+    async processQueue () {
+        if (!this.processing) {
+            while (this.taskQueue.length > 0) {
+                this.processing = true
+                const task = this.taskQueue.shift()
+                await task()
```

Comment:
should errors cause processing to halt for all?

This example wont run `console.log(3)`

```js

const Storage = new StorageInstance()
Storage.addTask(async () => {
    console.log('1')
})
Storage.addTask(async () => {
    console.log('2')
    throw new Error("oops!")
})
Storage.addTask(async () => {
    console.log('3')
    done()
})
```

Reply:
I think they should be rare enough to be safe to ignore, largely the tab state is safe to ignore a few writes to and potentially risk a bit of data loss.

I think the only likely cause is storage exceeded which we're currently not handling and likely will be a follow up to see if we should care / how to handle it :)